test/e2e/nodepool_upgrade_test.go

```diff
@@ -0,0 +1,114 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/releaseinfo"
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestUpgradeNodePool(t *testing.T) {
+	t.Parallel()
+	g := NewWithT(t)
+
+	ctx, cancel := context.WithCancel(testContext)
+	defer cancel()
+
+	t.Logf("Starting nodepool upgrade test from %s to %s", globalOpts.PreviousReleaseImage, globalOpts.LatestReleaseImage)
+
+	client, err := e2eutil.GetClient()
+	g.Expect(err).NotTo(HaveOccurred(), "failed to get k8s client")
+
+	clusterOpts := globalOpts.DefaultClusterOptions()
+	clusterOpts.ReleaseImage = globalOpts.LatestReleaseImage
+	clusterOpts.ControlPlaneAvailabilityPolicy = string(hyperv1.SingleReplica)
+
+	clusterOpts.BeforeApply = func(o crclient.Object) {
+		switch v := o.(type) {
```

Comment:
I'd expect this to be available in CreateNodePoolOptions and so then can be also exposed in the CLI.